Convert Tick Tank to use DeployToUpgrade + some fixes

penev92 opened the pull request:
Not much to describe here. PR is split into nice self-explanatory commits for your convenience.
~~Fixes an oversight from #9435.~~ (already fixed)
~~Fixes #9242.~~ (already fixed)
~~Fixes #9864.~~ (already fixed)
~~Fixes #9996.~~ (already fixed)

abcdefg30 commented:
> OpenRA.Utility(1,1): Error: Actor ttnk using voice set Vehicle does not define Action voice required by OpenRA.Mods.Common.Traits.AttackTurretedInfo

reaperrr commented:
You probably need to set `Voice: Attack` on `AttackTurreted`.

penev92 commented:
Fixed the voice.

atlimit8 commented:
`DeployingUpgrade` should not be active while deployed since deploying means getting deployed - not deployed.
You probably should add an `UndeployedUpgrade` to enable traits while fully undeployed. 

In fact, since `DeployedUpgrades` and `UndeployedUpgrade` could be used together to enable/disable traits while in transition as well, `DeployingUpgrade` could be entirely replaced by `UndeployedUpgrade`.

penev92 commented:
>DeployingUpgrade should not be active while deployed since deploying means getting deployed - not deployed.

True, but that makes for some ugly and redundant YAML, so I decided there wasn't any harm in keeping it.

>You probably should add an UndeployedUpgrade to enable traits while fully undeployed. 

That sounds like taking the whole "everything is an upgrade" idea to a whole new level, which doesn't sound like a something we'd want to start doing on a large scale.

atlimit8 commented:
If you are against adding `UndeployedUpgrade` then `DeployingUpgrade` should be renamed `NotUndeployedUpgrade` since "deploying" is incorrect.

On another note, "upgrades" should be renamed "conditions" this weekend, which should help with confusion on the matter especially when condition expression traits such as `ConditionCombiner` are added.

abcdefg30 commented:
Needs a rebase.

penev92 commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Activities/Attack.cs:
```diff
@@ -34,7 +34,7 @@ public Attack(Actor self, Target target, bool allowMovement, bool forceAttack)
 			this.forceAttack = forceAttack;
 
-			attack = self.Trait<AttackBase>();
+			attack = self.TraitsImplementing<AttackBase>().First(x => !x.IsTraitDisabled);
```
What if that specific AttackBase becomes disabled between the constructor and tick (which may be quite far apart if the activity is queued)?

atlimit8 replied:
Good point.
Perhaps `attack` should be initialized (`attack = self.TraitsImplementing<AttackBase>().FirstOrDefault(x => !x.IsTraitDisabled)`) on first `Tick` and the activity canceled if `attack == null`.

penev92 replied:
More like that should be fetched on every tick, which is a bit ridiculous.

atlimit8 replied:
Actually, all that needs to be done in other ticks is check if cached trait is disabled and cancel if it is.
```c#
 		public override Activity Tick(Actor self)
 		{
			if (attack == null)
			{
				attack = self.TraitsImplementing<AttackBase>().FirstOrDefault(x => !x.IsTraitDisabled)
				if (attack == null)
					return NextActivity;
			}
			else if (attack.IsTraitDisabled)
				return NextActivity;

 			var ret = InnerTick(self, attack);
 			attack.IsAttacking = ret == this;
 			return ret;
 		}
```

pchote replied:
If you store the IEnumerable<AttackBase> then the filtering each tick will be relatively cheap.

penev92 commented:
This now depends on #10014, so I'll close it for the moment.

penev92 committed:
Don't crash on multiple Attack* traits

This is important for actors that use upgrades to switch between attack types.

penev92 committed:
Convert the Tick Tank to use DeployToUpgrade

penev92 committed:
Remove now-redundant actor `gatick` (deployed tick tank)

penev92 committed:
Fix AttackFollow (and consequently AttackTurreted) firing at old target when the trait is reenabled

Most noticeable by ordering a deployed Tick Tank to fire at the ground, then undeploying it and deploying it again.

penev92 commented:
Revived.

abcdefg30 commented:
Looks good to me. :+1: 

But this needs a) a solution for the importer and b) a rebase.

penev92 committed:
Make the TS map importer aware of deployable actors who use upgrades instead of transforming

penev92 commented:
Done.

Mailaender commented:
The tick tank will stop firing when deploying/undeploying. This isn't a regression though.

Mailaender commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -100,12 +100,67 @@ TTNK:
-		NoTransformSounds:
-		Voice: Move
+	RenderSprites:
+		Image: ttnk
```
Isn't this redundant?

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c72403dfcee9a3e220625e1091c6f4686f35caa1...bfd4dc37658fb384985f4805a5042ac95713b342)
